RawKV Change Data Capture (#86)
tikv/rfcs: text/0083-rawkv-change-data-capture.md

# RawKV Change Data Capture

- RFC PR: https://github.com/tikv/rfcs/pull/86
- Tracking Issue: https://github.com/tikv/tikv/issues/11965

## Summary

This proposal introduces the technical design of **RawKV Change Data Capture** (abbr. **RawKV CDC**).

## Motivation

> Customers are deploying TiKV clusters as non-transactional key-value (**RawKV**) storage for application router or other scenarios. When **RawKV** being more and more widely adopted, [Change Data Capture] become a necessary feature to meet the high availability requirement of industries (*e.g.* banking and finance, to deploy disaster recovery clusters).

Reviewer: ```suggestion
Customers are deploying TiKV clusters as non-transactional key-value (**RawKV**) storage for application router or other scenarios. With **RawKV** being more and more widely adopted, [Change Data Capture] becomes a necessary feature to meet the high availability requirement of industries (*e.g.* banking and finance, to deploy disaster recovery clusters).
```

*(Now we will focus on the scenario of **Cross Cluster Replication**, i.e. replication between main and recovery TiKV clusters. Design for other scenarios will be supplemented later.)*

> In this proposal we introduces **TiKV-CDC**, a new component for **RawKV CDC**, which captures changed data from upstream/main TiKV cluster, and replicates to downstream/recovery TiKV cluster in real time.

Reviewer: ```suggestion
In this proposal we introduce **TiKV-CDC**, a new component for **RawKV CDC**, which captures changed data from upstream/main TiKV cluster, and replicates to downstream/recovery TiKV cluster in real time.
```

<img src="../media/rawkv-cdc.png" alt="RawKV CDC" width="60%"/>

## Key Challenges & Solutions

### Capturing Change

Timestamp is necessary for indicating which and when data is changed. So **RawKV CDC** will base on [TiKV API V2], which encodes timestamp with key.

### Consistency and Resolved Timestamp

Data in downstream cluster is expected to be consistent with upstream cluster. To be consistent:

**1st**, as **RawKV** does not provide transaction semantic, **RawKV CDC** will not guarantee the order of writes for different keys between upstream and downstream cluster.

**2nd**, [TiKV API V2] provides causality consistence, that if two writes (of the same key) `A` [Happened Before] `B`, then `Timestamp(A)` < `Timestamp(B)`. **RawKV CDC** guarantee the same causality consistence as well.

> **3rd**, [TiKV API V2] does not guarantee that a **RawKV** entry with smaller timestamp must be observed first. So **TiKV-CDC** will sort entires by timestamp before send to downstream.

Reviewer: ```suggestion
**3rd**, [TiKV API V2] does not guarantee that a **RawKV** entry with smaller timestamp must be observed first. So **TiKV-CDC** will sort entries by timestamp before send to downstream.
```

> But how **TiKV-CDC** knows it will not receive any even earlier entry from TiKV after the previous batch of data had been sent to downstream ? So we need **Resolved Timestamp** (abbr. **resolved-ts**). **Resolved-ts** is similar to [Watermark], by which TiKV indicating that all entries earlier have been observed. When **TiKV-CDC** receives a `resolved-ts`, all sorted entries with `timestamp <= resolved-ts` are safe to write to downstream.

Reviewer: ```suggestion
But how **TiKV-CDC** knows it will not receive any even earlier entry from TiKV after the previous batch of data had been sent to downstream ? So we need **Resolved Timestamp** (abbr. **resolved-ts**). **Resolved-ts** is similar to [Watermark], by which TiKV indicats that all entries earlier have been observed. When **TiKV-CDC** receives a `resolved-ts`, all sorted entries with `timestamp <= resolved-ts` are safe to write to downstream.
```

**Owner** is elected using [ETCD]. **Owner** and **Captures** are stateless, all metadata (mainly including tasks parameters and states) is stored in [ETCD]. So the availability and scalability are archived depending on [ETCD]. As size of metadata is small and update frequency is low, [ETCD] should be able to handle the workload.

For availability, we utilize a **Checkpoint** mechanism. **Checkpoint** is the latest **resolved-ts** that has been written to downstream successfully, and persist in [ETCD]. When **TiKV-CDC** cluster starts after planned or unexpected stop, replication tasks will resume from the **Checkpoint**.

**Checkpoint** will be affected by **Garbage Collection**. **Checkpoint** requires that all versions since **Checkpoint** are kept, otherwise **TiKV-CDC** would miss some changes. So **TiKV-CDC** will set a **Service Safe Point** on timestamp the same as **Checkpoint**, with **TTL** defaults to `24` hours. The **TTL** means that **TiKV-CDC** and replicate tasks should be resumed in no more than **TTL**, otherwise the **Checkpoint** will be invalid.

> Finally, in this proposal, one replication task will be processed in only one **Capture**, to make it right more easily. Splitting a task to sub-tasks will get better performance and scalability, but we must be caution to avoid any overlap or hole among sub-tasks, otherwise the correctness will be violated. Sub-tasks will be considered later.

Reviewer: ```suggestion
Finally, in this proposal, one replication task will be processed in only one **Capture**, to make it right more easily. Splitting a task to sub-tasks will get better performance and scalability, but we must be cautious to avoid any overlap or hole among sub-tasks, otherwise the correctness will be violated. Sub-tasks will be considered later.
```

### Recovery Point Objective

Customers require that [Recovery Point Objective] (abbr. **RPO**) should be no more than tens of seconds. In **RawKV CDC**, [RPO] is measured by the **Checkpoint Lag** metric, the interval between **Checkpoint** and **TSO** at the time.

Main factors affecting **Checkpoint Lag** includes:

- TiKV internal:
  - **Resolved-ts Lag**: the interval between **resolved-ts** and **TSO**. Which approximately equals to the longest duration of all RawKV writes at that moment. In most circumstances, it should be no more than several seconds.
- Data transmission on network:
  - **Network latency** cross cities would be `30` milliseconds or even more. But as we are using the high-performance [gRPC] as underlying implementation of networking, and writing to downstream by batch, we don't concern with network latency now.
- **TiKV-CDC** internal:

>   - **Sorter** would be the bottleneck, as it 1) process data from all regions, 2) holds too many data if **resolved-ts** or downstream is blocked, 3) sorts data in disks when data is too big to fit in memory.

Reviewer: ```suggestion
  - **Sorter** would be the bottleneck, as it 1) process data from all regions, 2) holds too much data if **resolved-ts** or downstream is blocked, 3) sorts data in disks when data is too big to fit in memory.
```

  - This issue will be even more severe on initial replication. In this scenario **Sorer** must hold all existing data from starting point of replication task to current before send to downstream.
  - To address this issue, we provide an alternative: using [TiKV-BR] to complete the initial replication. [TiKV-BR] backup and restore data fully utilizing power of the whole TiKV cluster, and don't need the sorting, so is much more efficient.

## Detailed Design

### TiKV

#### Resolved-ts

We will use [memory locks] to trace the "inflight" RawKV writes: acquire a lock before scheduling the writes, and release the lock after (asynchronously) send response to client.

> Then, CDC worker in TiKV is able to get the **resolved-ts** as `min(timestamp)-1` of all locks. If there is no lock, use current TSO as **resolved-ts**.

Reviewer: As concurrency manager [lock the ts before key write](https://github.com/tikv/tikv/blob/a80ab9880d6c25fe9b962e582846360dd229783b/src/storage/mod.rs#L1921-L1925). The `resolved-ts` use the lock ts is enough, no need `-1`.
